src/main/java/com/synacy/lesson04/exercise/service/BankAccountServiceImpl.java

```diff
@@ -36,17 +35,52 @@ public void withdraw(BankAccount bankAccount, BigDecimal amount)
 	@Override
 	public void deposit(BankAccount bankAccount, BigDecimal amount) {
 
+		BigDecimal currentBalance = bankAccount.getBalance();
+		BigDecimal newBalance = currentBalance.add(amount);
+
+		Transaction transaction = new Transaction(bankAccount, TransactionType.DEBIT, amount, new Date());
+		bankAccount.setBalance(newBalance);
+		transaction.setStatus(TransactionStatus.CLEARED);
+		transactionDao.saveTransaction(transaction);
+
+		bankAccountDao.saveBankAccount(bankAccount);
 	}
 
 	@Override
 	public void transfer(BankAccount sourceBankAccount, BankAccount destinationBankAccount, BigDecimal amount)
 			throws InsufficientBalanceException {
 
+		BigDecimal currentBalance = sourceBankAccount.getBalance();
+
+		if(currentBalance.compareTo(amount) < 0) {
+			throw new InsufficientBalanceException(sourceBankAccount, currentBalance, amount);
+		}
+
+		BigDecimal newBalance = currentBalance.subtract(amount);
+
+		Transaction creditTransaction = new Transaction(sourceBankAccount, TransactionType.CREDIT, amount, new Date());
+		sourceBankAccount.setBalance(newBalance);
+
+		currentBalance = destinationBankAccount.getBalance();
+		newBalance = currentBalance.add(amount);
```

Comment:
Refrain from using variables for different context. For this case, use two variables for calculating new balance for each bank account.